### DISCUSSION TITLE
Link prediction low performance at inference (hits@k)

### DISCUSSION THREAD

**Discussion by erenpolat** [pyg-team/pytorch_geometric#9356] [Q&A]
Hello, 

I've trained my dataset on a simple link prediction model and I am able to get >0.82 AUC score (downsampling the negative edges equal to the pos edges). When I want to use the same graph for inference (not even unseen graph, but the same big graph that I split to train/val/test), rank the probabilities and get the top k edges; performance is very low. Among the top 1000 edges, there are only 150 original edges. Among the first 50000, there are only 1000 original edges. What am I doing wrong here? (the graph has 2048 nodes and 2700 edges)

```
@torch.no_grad()
def infer_edges(pyg_graph, model):
    model.eval()
    z = model.encode(pyg_graph.x, pyg_graph.edge_index, pyg_graph.edge_attr)
    prob_adj = model.decode_all(z)
    print('prob_adj shape', prob_adj.shape)
    edges = [(i,j) for i in range(len(prob_adj[0])) for j in range(len(prob_adj[0]))]
    edge_to_prob = {}
    for e in edges:
        edge_to_prob[e] = prob_adj[e[0], e[1]].item()
    
    sorted_edges = sorted(edge_to_prob.items(), key=lambda item: item[1], reverse=True)

    original_edges = {}

    # Iterate over the columns of the edge_index tensor to extract edges
    for i in range(pyg_graph.edge_index.size(1)):
        source, target = pyg_graph.edge_index[0, i].item(), pyg_graph.edge_index[1, i].item()
        original_edges[(source, target)] = 0  # Initialize with 0 or any desired initial value

    k = 1000
    first_k = sorted_edges[:k]

    hits = 0    
    for e in first_k:
        if e[0] in original_edges:
            hits += 1

    print('hits: ', hits, 'original edges: ', pyg_graph.edge_index.size(1))
@torch.no_grad()
def get_val_loss(model, criterion, val_data):
    model.eval()
    z = model.encode(val_data.x, val_data.edge_index, val_data.edge_attr)

    # We perform a new round of negative sampling for every training epoch:
    neg_edge_index = negative_sampling(
        edge_index=val_data.edge_index, num_nodes=val_data.num_nodes,
        num_neg_samples=val_data.edge_label_index.size(1), method='sparse')

    edge_label_index = torch.cat(
        [val_data.edge_label_index, neg_edge_index],
        dim=-1,
    )
    edge_label = torch.cat([
        val_data.edge_label,
        val_data.edge_label.new_zeros(neg_edge_index.size(1))
    ], dim=0)

    out = model.decode(z, edge_label_index)

    return criterion(out, edge_label)

def train(model, optimizer, train_data, criterion):
    model.train()
    optimizer.zero_grad()
    z = model.encode(train_data.x, train_data.edge_index, train_data.edge_attr)

    # We perform a new round of negative sampling for every training epoch:
    neg_edge_index = negative_sampling(
        edge_index=train_data.edge_index, num_nodes=train_data.num_nodes,
        num_neg_samples=train_data.edge_label_index.size(1), method='sparse')


    edge_label_index = torch.cat(
        [train_data.edge_label_index, neg_edge_index],
        dim=-1,
    )
    edge_label = torch.cat([
        train_data.edge_label,
        train_data.edge_label.new_zeros(neg_edge_index.size(1))
    ], dim=0)

    out = model.decode(z, edge_label_index)
    print(out.shape)

    
    loss = criterion(out, edge_label)

    #div = pyg_divergence(pyg, flows, divs, out)
    
    loss.backward()
    optimizer.step()
    return loss, model


@torch.no_grad()
def test(model, data):
    model.eval()
    z = model.encode(data.x, data.edge_index, data.edge_attr)
    out = model.decode(z, data.edge_label_index).sigmoid()
    return roc_auc_score(data.edge_label.cpu().numpy(), out.cpu().numpy())

def main(): 
    device = 'cpu'
    exp = 'power'
    nonneg = exp == 'traffic'

    transform = T.Compose([
        T.ToDevice(device),
        T.RandomLinkSplit(num_val=0.1, num_test=0.2, is_undirected=True,
                        add_negative_train_samples=False, neg_sampling_ratio=1)
    ])

    pyg_graph = torch.load('power_wcoor_flow.pt')
    train_data, val_data, test_data = transform( pyg_graph)

    n_edge_features = pyg_graph.edge_attr.shape[-1]
    n_features = pyg_graph.x.shape[-1]


    model = Net(n_features, 32, 4).to(device)
    
    optimizer = torch.optim.Adam(params=model.parameters(), lr=0.01)
    criterion = torch.nn.BCEWithLogitsLoss()

    best_val_auc = final_test_auc = 0

    training_loss = []
    #train_aucs = []
    val_aucs = []
    val_losses = []
    
    best_model = None
    best_epoch = -1
    for epoch in range(1, 500):
        loss, model = train(model, optimizer, train_data, criterion)
        val_auc = test(model, val_data)

        val_loss = get_val_loss(model, criterion, val_data)
        val_losses.append(val_loss.item())

        val_aucs.append(val_auc)
        if val_auc > best_val_auc:
            best_val_auc = val_auc
            best_model = model
            best_epoch = epoch
        print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}, Val: {val_auc:.4f}')
    
    torch.save(best_model.state_dict(), 'link_pred_models/enyenisi3.pt')


    final_test_auc = test(best_model, test_data)
    print(f'Final Test: {final_test_auc:.4f}, Best epoch: {best_epoch}')

    plot_val(val_aucs, val_losses)
```

`class Net(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GCNConv(in_channels, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, out_channels)

    def encode(self, x, edge_index, edge_attr):
        x = self.conv1(x, edge_index).relu()
        return self.conv2(x, edge_index)

    def decode(self, z, edge_label_index):
        return (z[edge_label_index[0]] * z[edge_label_index[1]]).sum(dim=-1)

    def decode_all(self, z):
        prob_adj = z @ z.t()
        return prob_adj`

``

**Comment by rusty1s** [comment]
I think the issue is that you need to compute `hits@k` for each individual source node. Currently, it looks like you compute this globally.
